# Basic GPU testing

I explored the possibilty of combining gpu-acceleration with rasterizing polygons.

In [ ]:
%pip install -r requirements.txt

NotImplementedError: A UTF-8 locale is required. Got ANSI_X3.4-1968

In [ ]:
import geopandas as gpd
import dask.array as da
import xarray as xr
import datashader as ds
from datashader.transfer_functions import shade
from xrspatial import zonal_stats
from spatialpandas.geometry import PolygonArray
import spatialpandas as sp
import rioxarray

gdf = gpd.read_parquet('cagp_ml_data_feb_11_2025.parquet').set_crs(2272)
gdf = gdf[gdf.geometry.is_valid]
gdf = gdf.dropna(subset=['geometry'])
gdf['geometry']

0                            POINT (2673029.899 240341.196)
1                            POINT (2694650.973 249956.871)
2         POLYGON ((2700910.34 242074.708, 2700934.458 2...
3                            POINT (2704931.006 244870.908)
4                             POINT (2705992.43 243041.199)
                                ...                        
587044                       POINT (2693317.725 250451.995)
587045                       POINT (2698050.379 230481.905)
587046                       POINT (2697762.017 242387.067)
587047                       POINT (2692042.644 264020.095)
587048                         POINT (2686671.7 255636.899)
Name: geometry, Length: 583759, dtype: geometry

I decided to go with cupy as I could seamlessly integrate with xarray-spatial.

In [ ]:
import cupy as cp

In [ ]:
crs = 32618

In [ ]:
# Corrected Query for Sentinel-2 with Cloud Cover Filter
bbox = [-75.2803, 39.8670, -74.9557, 40.1379]  # Philadelphia bounding box
datetime = "2024-06-01/2024-08-31"  # Summer 2024
cloudy_less_than = 10  # Percent cloud cover threshold
filepath = 'sentinel_ndvi2.tif'

In [ ]:
polygons = sp.GeoDataFrame(gdf[['geometry']].to_crs(crs).representative_point().buffer(30), columns=['geometry'])
polygons['index'] = polygons.index.astype(int)

raster = rioxarray.open_rasterio(filepath, chunks="auto").squeeze()
raster = raster.rio.write_crs(crs)
xmin, ymin, xmax, ymax = raster.rio.bounds()
width, height = raster.shape
canvas = ds.Canvas(plot_width=width, plot_height=height,
                   x_range=(xmin, xmax), y_range=(ymin, ymax))

zones = canvas.polygons(polygons, geometry='geometry', agg=ds.any())
zones = zones.where(~zones.isnull(), 0)  # Convert NaN to 0

zones = canvas.polygons(polygons, geometry='geometry', agg=ds.max("index"))

In [ ]:
raster = raster.chunk({'x': 512, 'y': 512})

In [ ]:
zones = zones.chunk({'x': 512, 'y': 512})

In [ ]:
zones = zones.rio.write_crs(raster.rio.crs)  # Ensure CRS is identical
zones = zones.rio.reproject_match(raster)
zones.data = cp.asarray(zones.data)
raster.data = cp.asarray(raster.data)

In [ ]:
zones_cp, raster_cp= zones.compute(), raster.compute()

In [ ]:
from xrspatial.zonal import stats
# zones = zones.chunk({'x': 256, 'y': 256})
# zones = zones.rio.write_crs(raster.rio.crs)  # Ensure CRS is identical
# zones = zones.rio.reproject_match(raster)
res = stats(zones_cp, raster_cp, stats_funcs=['mean'])
res

zone      mean
0           33.0 -0.498259
1           49.0 -0.444909
2           50.0 -0.103618
3           52.0 -0.050891
4           59.0 -0.280519
...          ...       ...
167315  587044.0 -0.146434
167316  587045.0 -0.112433
167317  587046.0 -0.137555
167318  587047.0 -0.312132
167319  587048.0 -0.129814

[167320 rows x 2 columns]

167320 / 583759 $≈$ 0.2867 or 28.67% of the parcels had the valid data.

In [ ]:
res.to_parquet('ndvi_polygon_zonal.parquet')

# EE Data workflow

With that, I felt it was worth exploring how to scale with google earth engine integration.

In [ ]:
crs = 2272

Pulled the MAIP image from google earth engine with a 3m resolution and calibrated it accordingly.

In [ ]:
import geopandas as gpd
import dask.array as da
import xarray as xr
import datashader as ds
from datashader.transfer_functions import shade
from xrspatial import zonal_stats
from spatialpandas.geometry import PolygonArray
import spatialpandas as sp
import rioxarray

In [ ]:
import ee
import geemap
import xarray as xr
import rioxarray
ee.Authenticate()
ee.Initialize(project='ee-jeongjooho1995')

# Define a region of interest (ROI).
roi = ee.Geometry.Rectangle(bbox)

# Create a median composite image from a Landsat 8 image collection.
images = ee.ImageCollection("USDA/NAIP/DOQQ")\
         .filterBounds(roi)\
         .filterDate('2022-07-01', '2023-07-31')\
        .mosaic()
# token = "AIzaSyArgYJGHdk1CZgLRyUjwhSluvmj_pZ9AdQ"
ndvi = images.normalizedDifference(['N', 'R'])
# print(images)
# # Convert the Earth Engine image to an xarray DataArray.
# # The 'scale' parameter defines the spatial resolution in meters.
# geemap.ee_initialize(token)
da = geemap.ee_to_xarray(ndvi, crs=f"EPSG:{crs}", geometry=ee.Geometry.Rectangle(bbox), ee_initialize=False, scale=3)

da

/usr/local/lib/python3.11/dist-packages/xee/ext.py:688: UserWarning: Unable to retrieve 'system:time_start' values from an ImageCollection due to: No 'system:time_start' values found in the 'ImageCollection'.
  warnings.warn(


<xarray.Dataset> Size: 4GB
Dimensions:  (time: 1, X: 29318, Y: 33797)
Coordinates:
  * time     (time) int64 8B 0
  * X        (X) float64 235kB 2.662e+06 2.662e+06 ... 2.75e+06 2.75e+06
  * Y        (Y) float64 270kB 2.041e+05 2.041e+05 ... 3.055e+05 3.055e+05
Data variables:
    nd       (time, X, Y) float32 4GB ...
Attributes:
    crs:      EPSG:2272

In [ ]:

raster = da.nd.isel(time=0)
raster = raster.rio.write_crs(crs)
raster = raster.rename({'X': 'x', 'Y':'y'})
raster = raster.transpose(..., "y", "x")
raster


<xarray.DataArray 'nd' (y: 33797, x: 29318)> Size: 4GB
[990860446 values with dtype=float32]
Coordinates:
    time         int64 8B 0
  * x            (x) float64 235kB 2.662e+06 2.662e+06 ... 2.75e+06 2.75e+06
  * y            (y) float64 270kB 2.041e+05 2.041e+05 ... 3.055e+05 3.055e+05
    spatial_ref  int64 8B 0
Attributes:
    id:             nd
    data_type:      {'type': 'PixelType', 'precision': 'float', 'min': -1, 'm...
    crs_transform:  [1, 0, 0, 0, 1, 0]

Rasterized the polygon as before.

In [ ]:
polygons = sp.GeoDataFrame(gdf[['geometry']].to_crs(crs).representative_point().buffer(30), columns=['geometry'])
polygons['index'] = polygons.index.astype(int)


xmin, ymin, xmax, ymax = raster.rio.bounds()
width, height = raster.shape
canvas = ds.Canvas(plot_width=width, plot_height=height,
                   x_range=(xmin, xmax), y_range=(ymin, ymax))

zones = canvas.polygons(polygons, geometry='geometry', agg=ds.any())
zones = zones.where(~zones.isnull(), 0)  # Convert NaN to 0

zones = canvas.polygons(polygons, geometry='geometry', agg=ds.max("index"))

In [ ]:
zones = zones.rio.write_crs(raster.rio.crs)  # Ensure CRS is identical
zones = zones.rio.reproject_match(raster)

I tried to use the cupy arrays themselves but the raster was simply too large for the GPU to handle.

In [ ]:
res = stats(zones.data, raster.data, stats_funcs=['mean'])
res

So, I decided to attach the cuda backend for dask to distribute the load.

## Limitations
Unfortunately, xarray-spatial does not support cupy-backed dask array for zonal statistics. I only discovered it after trying it out and took a closer look at the documentation on github.
https://github.com/makepath/xarray-spatial

In [ ]:
zones.data = cp.asarray(zones.data)

In [ ]:
raster = raster.chunk({'x': 512, 'y': 512})

In [ ]:
zones = zones.chunk({'x': 512, 'y': 512})

In [ ]:
raster.data = cp.asarray(raster.data)

In [ ]:
type(zones.data)

dask.array.core.Array

In [ ]:
from xrspatial.zonal import stats


res = stats(zones, raster, stats_funcs=['mean'])
res

Dask DataFrame Structure:
                  zone     mean
npartitions=1                  
               float64  float64
                   ...      ...
Dask Name: getitem, 8 expressions
Expr=(ColumnsSetter(frame=Concat(frames=[FromGraph(0c11df3), FromGraph(a72f06c), FromGraph(9ff0c90), FromGraph(70cd5f6), FromGraph(1d9de90)], axis=1), columns=dict_keys(['zone', 'count', 'sum', 'sum_squares', 'mean'])))[['zone', 'mean']]

In [ ]:
res.to_parquet('ndvi_polygon_zonal.parquet')

KeyboardInterrupt: 

## Recombined image Attempt

With Nissim's advice, I settled on using recombined images from sentinel + NAIP and computing the cupy array.

In [ ]:
!pip install -r requirements.txt

INFO: pip is looking at multiple versions of spatialpandas to determine which version is compatible with other requirements. This could take a while.
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 89.4/89.4 kB 7.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 194.2/194.2 kB 16.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 61.9/61.9 kB 5.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 99.6/99.6 kB 8.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 22.2/22.2 MB 85.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.1/2.1 MB 75.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 123.9/123.9 kB 11.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 18.3/18.3 MB 97.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.0/2.0 MB 77.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.3/1.3 MB 66.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 155.1/155.1 kB 13.9 M

In [ ]:
import ee
import geemap
import xarray as xr
import rioxarray
import geopandas as gpd
import dask.array as da
import xarray as xr
import datashader as ds
from datashader.transfer_functions import shade
from xrspatial import zonal_stats
from spatialpandas.geometry import PolygonArray
import spatialpandas as sp
import rioxarray

In [ ]:
ee.Authenticate()
ee.Initialize(project='ee-jeongjooho1995')

In [ ]:
# Corrected Query for Sentinel-2 with Cloud Cover Filter
bbox = [-75.2803, 39.8670, -74.9557, 40.1379]  # Philadelphia bounding box
datetime = "2024-06-01/2024-08-31"  # Summer 2024
cloudy_less_than = 10  # Percent cloud cover threshold
crs = 32618

In [ ]:
gdf = gpd.read_parquet('cagp_ml_data_feb_11_2025.parquet').set_crs(2272)
gdf = gdf[gdf.geometry.is_valid]
gdf = gdf.dropna(subset=['geometry'])
gdf['geometry']

0                            POINT (2673029.899 240341.196)
1                            POINT (2694650.973 249956.871)
2         POLYGON ((2700910.34 242074.708, 2700934.458 2...
3                            POINT (2704931.006 244870.908)
4                             POINT (2705992.43 243041.199)
                                ...                        
587044                       POINT (2693317.725 250451.995)
587045                       POINT (2698050.379 230481.905)
587046                       POINT (2697762.017 242387.067)
587047                       POINT (2692042.644 264020.095)
587048                         POINT (2686671.7 255636.899)
Name: geometry, Length: 583759, dtype: geometry

In [ ]:
phillyBox = ee.Geometry.Rectangle(bbox)

roi = ee.Geometry.Rectangle(bbox)
def f(image):
    swirDown = image.select('B11') \
                    .resample('bilinear') \
                    .reproject(crs=image.select('B4').projection())
    return image.addBands(swirDown, None, True)

s2 = ee.ImageCollection("COPERNICUS/S2_SR_HARMONIZED")\
         .filterBounds(roi)\
         .filterDate('2024-06-01', '2024-08-31')\
         .filter(ee.Filter.lt('CLOUDY_PIXEL_PERCENTAGE', cloudy_less_than))\
         .map(f)\
        .mosaic()
ndvi = s2.normalizedDifference(['B8', 'B4'])
naip = ee.ImageCollection('USDA/NAIP/DOQQ')\
  .filterBounds(phillyBox)\
  .filter(ee.Filter.date('2022-06-01', '2022-08-31'))\
  .mosaic()

ndvi = ndvi.updateMask(ndvi.gte(-2))

ndvi5m = ndvi.reproject(
  crs= naip.projection(),
  scale=5
)
naipRGB = naip.select(['R', 'G', 'B'])


ndvi5m = ndvi5m.updateMask(ndvi5m.gte(-2))

intensity = naipRGB.reduce(ee.Reducer.mean()).rename('intensity')
edge = intensity.gradient().abs().reduce(ee.Reducer.sum()).rename('edge')


edgeNorm = edge.unitScale(0, edge.reduceRegion(
  reducer= ee.Reducer.percentile([98]),
  geometry= phillyBox,
  scale= 10,
  maxPixels= 1e9
).values().get(0))

fusedNDVI = ndvi5m.add(
  edgeNorm.multiply(0.1).multiply(ndvi5m.abs())
)

In [ ]:
crs=2272

In [ ]:
raster = geemap.ee_to_xarray(fusedNDVI, crs=f"EPSG:{crs}", geometry=ee.Geometry.Rectangle(bbox), ee_initialize=False, scale=5)
raster = raster.nd.isel(time=0)

/usr/local/lib/python3.11/dist-packages/xee/ext.py:688: UserWarning: Unable to retrieve 'system:time_start' values from an ImageCollection due to: No 'system:time_start' values found in the 'ImageCollection'.
  warnings.warn(


In [ ]:
raster = raster.rename({'X':'x', 'Y':'y'})
raster = raster.transpose('y', 'x')
raster = raster.chunk({"x": 1024, "y": 1024})



In [ ]:
polygons = sp.GeoDataFrame(gdf[['geometry']].to_crs(crs).representative_point().buffer(30), columns=['geometry'])
polygons['index'] = polygons.index.astype(int)


xmin, ymin, xmax, ymax = raster.rio.bounds()
width, height = raster.shape
canvas = ds.Canvas(plot_width=width, plot_height=height,
                   x_range=(xmin, xmax), y_range=(ymin, ymax))

zones = canvas.polygons(polygons, geometry='geometry', agg=ds.any())
zones = zones.where(~zones.isnull(), 0)  # Convert NaN to 0

zones = canvas.polygons(polygons, geometry='geometry', agg=ds.max("index"))

In [ ]:
zones = zones.rio.write_crs(raster.rio.crs)  # Ensure CRS is identical
zones = zones.rio.reproject_match(raster)

In [ ]:
zones = zones.chunk({'x': 1024, 'y': 1024})

In [ ]:
import cupy as cp

In [ ]:
zones.data = cp.asarray(zones.data)

In [ ]:
raster.data = cp.asarray(raster.data)

In [ ]:
zones_cp = zones.compute()
raster_cp = raster.compute()

In [ ]:
from xrspatial.zonal import stats

res = stats(zones_cp, raster_cp, stats_funcs=['mean'])

In [ ]:
res.to_parquet('ndvi_polygon_zonal2.parquet')

I've manage to get about 267433 / 583759 $≈45.8$% of the zones' raster computations with the mixed images.

In [ ]:
res

zone      mean
0            2.0  0.024696
1            6.0  0.177722
2            7.0  0.338136
3           10.0  0.428059
4           11.0  0.435190
...          ...       ...
267428  587041.0  0.438420
267429  587043.0  0.177677
267430  587044.0  0.139593
267431  587045.0  0.096738
267432  587046.0  0.103635

[267433 rows x 2 columns]